File: Funciones Recursivas/NQueens.cpp
#include <iostream>
#include <math.h>

using namespace std;

int c = 1;

bool Comprobar(int Queens[], int n, int k){
    for(int i = 0; i < k; i++)
        if((Queens[k] == Queens[i]) || (abs(k-i) == abs(Queens[k] - Queens[i])))
            return false;
    return true;
}

void NQueens(int Queens[], int n, int k){
    if(n==k){
        cout << "Solucion "<<c<<": ";
        for(int i = 0; i<n; i++)
            cout << Queens[i] << ' ';
        cout << endl;
        c++;
        return;
    }

    for(int i = 0; i < n; i++){
        Queens[k] = i;
        if(Comprobar(Queens, n, k))
            NQueens(Queens, n, k+1); 
    }

}

int main(){
    int N = 12;
    int *Queens = new int(N);
    NQueens(Queens, N, 0);
}

File: S06_Algorithms/NQueens.py
c = 1

def Comprobar(Queens, n, k):
    for i in range(k):
        if Queens[i] == Queens[k] or abs(Queens[i]- Queens[k]) == abs(i-k):
            return False
    return True

def NQueens(Queens, n, k):
    global c
    if k == n:
        print('Solución',c,':',Queens)
        c += 1
        return

    for i in range(n):
        Queens[k] = i
        if Comprobar(Queens,n,k):
            NQueens(Queens,n,k+1)


N = 12
Queens = N*[-1]
NQueens(Queens, N, 0)